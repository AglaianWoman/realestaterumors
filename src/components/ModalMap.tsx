
import GoogleMapReact from 'google-map-react'
import * as React from 'react'
import { connect } from 'react-redux'
import { ReduxState, ReduxStateMapbox } from '../reducer'
import { Actions as A } from '../reduxActions'

import * as Modal from 'antd/lib/modal'
import 'antd/lib/modal/style/css'

import * as InputNumber from 'antd/lib/input-number'
import 'antd/lib/input-number/style/css'


interface ModalMapProps {
  longitude: number
  latitude: number
  showModal: boolean
}
interface ModalMapState {
  mapWidth: string | number
  mapHeight: string | number
}


export class ModalMap extends React.Component<ModalMapProps, any> {

  state = {
    mapWidth: 600,
    mapHeight: 400,
  }

  static defaultProps = {
    longitude: 153,
    latitude: -27,
  }

  componentDidMount() {
    window.addEventListener('resize', this.handleWindowResize)
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.handleWindowResize)
  }

  handleWindowResize = () => {
    this.setState({
      mapWidth: window.innerWidth * 0.8
    })
  }

  showModal = () => {
    this.props.toggleShowModal(true)
  }

  handleOk = () => {
    console.log("Ok. Unmounting GoogleMapReact component.")
    this.props.toggleShowModal(false)
  }

  handleCancel = () => {
    console.log("Cancelling. Unmounting GoogleMapReact component.")
    this.props.toggleShowModal(false)
  }


  initStreetView = ({ map, maps }: { map: google.maps.MapOptions, maps: google.maps }) => {

    let sv: google.maps.StreetViewService = new maps.StreetViewService()
    let panorama: google.maps.StreetViewPanorama = new maps.StreetViewPanorama(document.getElementById('modalmap'))

    const processSVData = (data: google.maps.StreetViewPanoramaData, status: google.maps.StreetViewStatus) => {
      var marker: google.maps.Marker = new maps.Marker({
        position: data.location.latLng,
        title: data.location.description,
        map: map,
      })
      panorama.setPano(data.location.pano)
      panorama.setPov({ heading: 200, pitch: 0 })
      panorama.setVisible(true)
    }

    let svLocationRequest: google.maps.StreetViewLocationRequest = {
      location: {
        lat: this.props.latitude,
        lng: this.props.longitude,
        radius: 50
      }
    }
    sv.getPanorama(svLocationRequest, processSVData)
  }

  render() {
    return (
      <div>
        <button id='antd-login' onClick={this.showModal}>
          Open Modal Map
        </button>

        <Modal title='Street View' visible={this.props.showModal}
          width={window.innerWidth * 0.84}
          onOk={this.handleOk} onCancel={this.handleCancel}
        >
          <span>Longitude: {this.props.longitude}</span>
          <pre/>
          <span>Latitude: {this.props.latitude}</span>
          <pre/>
          {(
            this.props.showModal &&
            <div id='modalmap' style={{ width: this.state.mapWidth, height: this.state.mapHeight }}>
              <GoogleMapReact
                bootstrapURLKeys={{key: "AIzaSyDoEtrs7w3fIHSDvbPB7sAUw7tY7bIuAAU"}}
                onGoogleApiLoaded={this.initStreetView}
                options={{ streetViewControl: true }}
                yesIWantToUseGoogleMapApiInternals={true}
                defaultCenter={{ lat: this.props.latitude, lng: this.props.longitude }}
                defaultZoom={12} >
              </GoogleMapReact>
            </div>
          )}
        </Modal>
      </div>
    )
  }

}



const mapStateToProps = ( state: ReduxState ): ReduxStateMapbox => {
  return {
    showModal: state.reduxMapbox.showModal
  }
}

const mapDispatchToProps = ( dispatch ) => {
  return {
    toggleShowModal: (bool: boolean) => dispatch(
      { type: A.Mapbox.SHOW_MODAL, payload: bool }
    )
  }
}


export default connect(mapStateToProps, mapDispatchToProps)( ModalMap )




