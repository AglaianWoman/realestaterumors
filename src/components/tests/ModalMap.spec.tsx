

import * as React from 'react'
import * as renderer from 'react-test-renderer'
import { mount, shallow } from 'enzyme'

import { ModalMap } from '../ModalMap'
import * as Button from 'antd/lib/button'
import 'antd/lib/button/style/css'


let MockProps = {
  longitude: 0,
  latitude: 0,
}


test('<ModalMap/> component should contain 1 <Button> to toggle Modal', () => {
  const el = shallow( <ModalMap {...MockProps} /> )
  expect(el.find(Button).length).toEqual(1)
})


test('<ModalMap/> -> <Button> -> <span> should contain text: "Open Modal Map".', () => {
  const el = shallow( <ModalMap {...MockProps} /> )
  expect(el.find(Button).first().childAt(0).text()).toEqual('Open Modal Map')
})


test('<ModalMap/> should contain div: "#modalmap" when visible state is true', () => {
  const el = shallow( <ModalMap {...MockProps} /> )
  expect(el.find('#modalmap').length).toEqual(0)
  el.setProps({ showModal: true })
  expect(el.find('#modalmap').length).toEqual(1)
})



test('<ModalMap/> component should contain 2 <InputNumber> elements for lat and lng', () => {
  const el = shallow( <ModalMap {...MockProps} /> )
  expect(el.find('span').length).toEqual(2)
})


test('<ModalMap/> component matches snapshot', () => {
  const elR = renderer.create( <ModalMap {...MockProps} /> )
  expect(elR.toJSON()).toMatchSnapshot()
})






