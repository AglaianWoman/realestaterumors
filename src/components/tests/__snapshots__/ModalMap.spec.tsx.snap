exports[`test <ModalMap/> component matches snapshot 1`] = `
<div>
  <button
    className="ant-btn ant-btn-primary"
    onClick={[Function]}
    onMouseUp={[Function]}
    type="button">
    <span>
      Open Modal Map
    </span>
  </button>
</div>
`;
